### PR TITLE
feat(polls): handle events from ruma, maintain state

### PR DESCRIPTION
**Pull request by csmith**
This updates the Ruma dependency to include support for unstable
MSC3381v1 polls, handles the relevant events, and plumbs them
into the FFI layer.

Poll vote and end events update the corresponding start event in
the timeline. We keep track of all the votes we've seen for a
poll as this is required to aggregate them later (we can't, for
instance, drop a user's previous votes as we may then get a
poll end event with a timestamp between the two votes, which
would invalidate the more recent one).

It purposely does not handle:
 - Aggregating votes: will be done in a later PR
 - Editing polls: will be done in the future
 - Votes/End events received when we don't have the start event:
   will be done in a later PR

---

What needs to happen for this to not be a draft:

- [x] [The ruma PR](https://github.com/ruma/ruma/pull/1589) lands
- [X] [The first PR to add FFI types](https://github.com/matrix-org/matrix-rust-sdk/pull/2326) should be merged, as this is based on top of it
- [x] Cargo.toml should be updated to point at the normal ruma repo with the right commit

### CODE REVIEW

File: bindings/matrix-sdk-ffi/Cargo.toml [hunk 1]
```diff
@@ -29,7 +29,7 @@ eyeball-im = { workspace = true }
 extension-trait = "1.0.1"
 futures-core = { workspace = true }
 futures-util = { workspace = true }
-matrix-sdk-ui = { path = "../../crates/matrix-sdk-ui", default-features = false, features = ["e2e-encryption", "experimental-room-list", "experimental-encryption-sync", "experimental-notification-client", "experimental-sync-service"] }
+matrix-sdk-ui = { path = "../../crates/matrix-sdk-ui", default-features = false, features = ["e2e-encryption", "experimental-polls", "experimental-room-list", "experimental-encryption-sync", "experimental-notification-client", "experimental-sync-service"] }
```
Comment (Hywan) [inline bindings/matrix-sdk-ffi/Cargo.toml:32]: New: We consider that everything inside `matrix-sdk-ui` is experimental.
Consequence: Please remove the `experimental-polls` flag :-).

File: bindings/matrix-sdk-ffi/src/timeline.rs [hunk 2]
```diff
@@ -1252,3 +1286,15 @@ impl From<&matrix_sdk_ui::timeline::AnyOtherFullStateEventContent> for OtherStat
...
+pub enum PollKind {
+    Disclosed,
+    Undisclosed,
+}
+
+#[derive(Clone, uniffi::Record)]
+pub struct PollAnswer {
+    pub id: String,
+    pub text: String,
+}
```
Comment (Hywan) [inline bindings/matrix-sdk-ffi/src/timeline.rs:1300]: Is it the code from https://github.com/matrix-org/matrix-rust-sdk/pull/2326? Can we close #2326 in favor of this PR?

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs [hunk 3]
```diff
@@ -309,6 +318,18 @@ impl<'a> TimelineEventHandler<'a> {
                 AnyMessageLikeEventContent::Sticker(content) => {
                     self.add(should_add, TimelineItemContent::Sticker(Sticker { content }));
                 }
+                #[cfg(feature = "experimental-polls")]
+                AnyMessageLikeEventContent::UnstablePollStart(c) => {
+                    self.add( should_add, TimelineItemContent::Poll(PollState::from(c.poll_start)));
```
Comment (Hywan) [inline crates/matrix-sdk-ui/src/timeline/event_handler.rs:323]:
```suggestion
                    self.add(should_add, TimelineItemContent::Poll(PollState::from(c.poll_start)));
```

File: bindings/matrix-sdk-ffi/src/timeline.rs (hunk 2 above)
Comment (jplatte) [reply bindings/matrix-sdk-ffi/src/timeline.rs:1300]: See PR description (in particular todo list at the end) 😉

Comment (Hywan) [reply bindings/matrix-sdk-ffi/src/timeline.rs:1300]: /me whistles

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs (hunk 3 above)
Comment (csmith) [reply crates/matrix-sdk-ui/src/timeline/event_handler.rs]: Done!

File: bindings/matrix-sdk-ffi/Cargo.toml (hunk 1 above)
Comment (csmith) [reply bindings/matrix-sdk-ffi/Cargo.toml]: Neat! Done.

File: crates/matrix-sdk-ui/Cargo.toml [hunk 4]
```diff
@@ -28,13 +28,13 @@ futures-util = { workspace = true }
 imbl = { version = "2.0.0", features = ["serde"] }
 indexmap = "2.0.0"
 itertools = { workspace = true }
-matrix-sdk = { version = "0.6.2", path = "../matrix-sdk", default-features = false, features = ["experimental-sliding-sync"] }
+matrix-sdk = { version = "0.6.2", path = "../matrix-sdk", default-features = false }
```
Comment (csmith) [inline crates/matrix-sdk-ui/Cargo.toml:31]: Not sure how I managed that, will revert...

**Comment by zecakeh** [comment]
The Ruma PR has landed :slightly_smiling_face:

**Comment by julioromano** [comment]
Bump ruma PR here: https://github.com/matrix-org/matrix-rust-sdk/pull/2344

**Comment by julioromano** [comment]
Superseded by https://github.com/matrix-org/matrix-rust-sdk/pull/2381

**Pull request closed without merging**